### ISSUE TITLE
Certain MP3 files not correctly identified, and as a result, they are unplayable in a game.

### ISSUE THREAD

**Issue by Enjay001** [libsndfile/libsndfile#1039]
This issue was reported on the ZDoom forums.
https://forum.zdoom.org/viewtopic.php?t=79676

I found some MP3 files that GZDoom cannot play.  GZDoom uses libsndfile.

As can be seen from the discussion on the above linked forum page, for some reason, the files have some empty data before the music data actually starts.  This seems to be the cause of libsndfile being unable to identify the files as MP3s and, as a result, GZDoom cannot play them.

Media players play the files, and GZDoom used to.  So, even though they are unusual, the files are playable.

I have attached the two MP3 files discussed in the ZDoom forum below.

[MusicSamples.zip](https://github.com/user-attachments/files/17423174/MusicSamples.zip)

**Comment by kcat** [comment]
Seems it might be due to `mpeg_decoder_init` explicitly setting the `MPG123_NO_FRANKENSTEIN` flag, making libmpg123 more strict with the bytestream and causing it to reject the files since they start with 0 data and junk.

**Comment by kcat** [comment]
I've done a bit more testing, and it doesn't seem that `MPG123_NO_FRANKENSTEIN` is the culprit. Using libmpg123 directly myself I can open and decode the problem files just fine, and it still works when adjusting the flags as libsndfile does. But libsndfile fails to open the files all the same (other MP3 files I have do work with libsndfile though).

Maybe libsndfile is failing earlier and not getting to the mpeg decoder at all. Is there some environment variable or something to enable the logging the library does?

**Comment by fabiangreffrath** [comment]
Indeed, the `sndfile-play` application is able to play back these files by guessing from their file extensions that they might contain mpeg streams and handing them over to mpg123. However, once the file extensions are changed or removed, it falls flat on its face. 

https://github.com/libsndfile/libsndfile/blob/52b803f57a1f4d23471f5c5f77e1a21e0721ea0e/src/sndfile.c#L2755-L2761

**Comment by Enjay001** [comment]
I don't know if there has been any attempt to address this, or perhaps it simply can't be addressed, but I stumbled across the test files on my web space and was going to delete them.

I actually couldn't even remember what they were for.  So, I tested them quickly in the latest UZDoom Nightly build and the problem still exists for UZDoom and the test files.

That may well be as expected, but I thought I'd mention it just in case.

**Comment by fabiangreffrath** [comment]
Woof now has embedded the minimp3 library as a fallback for music lumps that pass through libsndfile's format detection. I have read somewhere that you can't tell that some file is a valid MP3 file before you actually attempt to play it back. 🤷

**Comment by kcat** [comment]
The issue tends to be MP3s having some junk data before the actual MP3 data stream starts. I think MP3 is generally handled is by checking the first N bytes (128? 1024?) for a valid frame header, rather than checking that the data stream contains a valid frame header immediately at the start. And MP3 is checked later in the detection order to avoid preempting other formats that can be more reliably detected.

Detecting a valid frame header can be followed by extra validation to reduce false positives, for example that a second frame follows with sensible parameter continuity. If a valid frame header is found, but following that frame is invalid junk data, it was a fluke and can return failure for mp3 detection and try another format. Or if a second frame is found, but the various parameters for bitrate, sample rate, channel count, etc, are unexpectedly different, it's another detection failure. It can also use the presence of an ID3v2 tag (or ID3v1, if random access is possible) as a hint for likely being mp3.
